src/runtime/gc_list.go

```diff
@@ -0,0 +1,460 @@
+		// There is no suitable allocation.
+		return nil
+	}
+
+	mem := (*allocHeader)(unsafe.Pointer(bestStart))
+	*bestDst, mem.next = mem, *bestDst
+	mem.len = size
+	return mem
+}
+
+func free(ptr unsafe.Pointer) {
+	// TODO: free allocations on request, when the compiler knows they're unused.
+}
+
+// GC performs a garbage collection cycle.
+func GC() {
+	if gcAsserts {
+		gcRunning = true
+	}
+
+	if gcDebug {
+		println("running collection cycle...")
+	}
+
+	// Mark in-use allocations.
+	remaining := mark()
+
+	if gcDebug {
+		println("found unreferenced allocations:")
+		activeAllocs.dump()
+	}
+
+	// Update the active allocations list.
+	activeAllocs = remaining
+
+	if gcDebug {
+		println("sorting active allocations")
+	}
+
+	activeAllocs.sort()
+
+	if gcDebug {
+		println("garbage collection cycle complete")
+	}
+
+	if gcAsserts {
+		gcRunning = false
+	}
+}
+
+// mark all in-use allocations and save them as a sorted list.
+func mark() allocList {
+	if gcDebug {
+		println("running GC mark phase...")
+		println("pre-mark allocations:")
+		activeAllocs.dump()
+	}
+
+	// Mark all allocations referenced by currently-executing functions.
+	markStack()
+
+	// Mark all allocations referenced by global variables.
+	markGlobals()
+
+	// Save remaining allocations to a list after they have been marked and scanned.
+	var remainingAllocs allocList
+
+	// Channel operations in interrupts may move task pointers around while we are marking.
+	// Therefore we need to scan the runqueue seperately.
+	var markedTaskQueue task.Queue
+runqueueScan:
+	if baremetal && hasScheduler {
+		for !runqueue.Empty() {
+			// Pop the next task off of the runqueue.
+			t := runqueue.Pop()
+
+			// Mark the task if it has not already been marked.
+			markRoot(uintptr(unsafe.Pointer(&runqueue)), uintptr(unsafe.Pointer(t)))
+
+			// Push the task onto our temporary queue.
+			markedTaskQueue.Push(t)
+		}
+	}
+
+	// Mark allocations referenced by other marked allocations.
+	for {
+		// Pop an allocation off of the scan list.
+		alloc := scanList.pop()
+		if alloc == nil {
+			break
+		}
+
+		// Scan the allocation.
+		alloc.scan()
+
+		// Save the allocation in the remaining allocations list.
+		remainingAllocs.push(alloc)
+	}
+
+	if baremetal && hasScheduler {
+		// Restore the runqueue.
+		i := interrupt.Disable()
+		if !runqueue.Empty() {
+			// Something new came in while finishing the mark.
+			interrupt.Restore(i)
+			if gcDebug {
+				println("scanning tasks queued during GC")
+			}
+			goto runqueueScan
+		}
+		runqueue = markedTaskQueue
+		interrupt.Restore(i)
+	}
+
+	if gcDebug {
+		println("GC mark phase complete")
+		println("remaining allocations:")
+		remainingAllocs.dump()
+	}
+
+	return remainingAllocs
+}
+
+// sort the allocation list in ascending address order.
+func (list *allocList) sort() {
+	// Sort the list by repeatedly moving the lowest-address element to the front.
+	// This is selection sort (https://en.wikipedia.org/wiki/Selection_sort).
```

Comment:
I'm slightly worried that a quadratic algorithm in a GC might be a problem. Maybe not on the AVR though with its very limited memory.

I think there is a way to avoid this sorting, at least if it's fine to limit the maximum object size to half the address space (which seems like a reasonable limitation to me). You can use one bit of `allocHeader.len` (e.g. the top bit) to store the mark state. Then, during sweeping, you can iterate through all objects and remove the ones that haven't been marked.

Reply:
That seems reasonable.

Reply:
It's just an idea for how it can be improved. The code as-is already looks good to me.

Reply:
Actually, I think this may be better as-is. That tagging strategy would actually also make the scan quadratic (the list needs to be fully traversed each time we need to find what to scan). I think the current strategy with the scan stack is at least a bit more readable.